File: combinatorics/resource/ST/project_5/bai_toan_4/trees/first_child_next_sibling/fcns_to_ap.cpp
#include <iostream>
#include <vector>
using namespace std;

void dfs(int u, int p, const vector<int>& fc, const vector<int>& ns, vector<int>& parent) {
    parent[u] = p;
    if (fc[u] != -1)
        dfs(fc[u], u, fc, ns, parent);
    if (ns[u] != -1)
        dfs(ns[u], p, fc, ns, parent);
}

int main() {
    int n;
    cout << "Nhap so nut: ";
    cin >> n;

    vector<int> fc(n), ns(n), parent(n, -1);
    cout << "Nhap mang fc:\n";
    for (int i = 0; i < n; ++i) cin >> fc[i];
    cout << "Nhap mang ns:\n";
    for (int i = 0; i < n; ++i) cin >> ns[i];

    dfs(0, -1, fc, ns, parent); // Bat dau tu goc 0

    cout << "Mang parent:\n";
    for (int i = 0; i < n; ++i)
        cout << parent[i] << " ";
    cout << "\n";

    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_4/trees/first_child_next_sibling/fcns_to_ap.py
def dfs(u, p, fc, ns, parent):
    parent[u] = p  # Gan cha cho nut u
    if fc[u] != -1:
        dfs(fc[u], u, fc, ns, parent)
    if ns[u] != -1:
        dfs(ns[u], p, fc, ns, parent)  # Anh em giu cha giong nhau

n = int(input("Nhap so nut: "))
print("Nhap mang fc:")
fc = list(map(int, input().split()))
print("Nhap mang ns:")
ns = list(map(int, input().split()))

parent = [-1] * n  # Khoi tao mang cha

dfs(0, -1, fc, ns, parent)  # Goc la 0

print("Mang parent:")
print(parent)